zcncore/transaction_query_base.go

```diff
@@ -457,6 +406,78 @@ func (tq *TransactionQuery) getFastConfirmation(ctx context.Context, txnHash str
 	return nil, nil, nil, thrown.Throw(ErrTransactionNotFound, strconv.Itoa(result.StatusCode))
 }
 
+func (tq *TransactionQuery) GetInfo(ctx context.Context, query string) (*QueryResult, error) {
+
+	consensuses := make(map[int]int)
+	var maxConsensus int
+	var consensusesResp QueryResult
+	// {host}{query}
+	err := tq.FromAll(ctx, query,
+		func(qr QueryResult) bool {
+			//ignore response if it is network error
+			if qr.StatusCode >= 500 {
+				return false
+			}
+
+			consensuses[qr.StatusCode]++
+			if consensuses[qr.StatusCode] > maxConsensus {
+				maxConsensus = consensuses[qr.StatusCode]
+				consensusesResp = qr
+			}
+
+			// If number of 200's is equal to number of some other status codes, use 200's.
+			if qr.StatusCode == http.StatusOK && consensuses[qr.StatusCode] == maxConsensus {
+				maxConsensus = consensuses[qr.StatusCode]
+				consensusesResp = qr
+
+				// consensus has been reached, don't waiting for other requests
+				if maxConsensus*100/tq.max >= consensusThresh {
+					return true
+				}
+			}
+
+			return false
+
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if maxConsensus == 0 {
+		return nil, stderrors.New("zcn: query not found")
+	}
+
+	rate := maxConsensus * 100 / tq.max
+	if rate < consensusThresh {
+		return nil, ErrInvalidConsensus
+	}
+
+	if consensusesResp.StatusCode != http.StatusOK {
+		return nil, stderrors.New(string(consensusesResp.Content))
+	}
+
+	return &consensusesResp, nil
+}
+
+func MakeSCRestAPICall(scAddress string, relativePath string, params map[string]string) ([]byte, error) {
+
+	path := fmt.Sprintf("/v1/screst/%v/%v", scAddress, relativePath)
+	query := withParams(path, Params(params))
+
+	tq, err := NewTransactionQuery(util.Shuffle(_config.chain.Sharders))
```

Comment:
Will this send requests to all the sharders? if so it may cause requests storm in large network.

Reply:
@cnlangzi let's make this requests go to 30% of the sharders rather than all sharders, also send to min 3 and max 10 to avoid `total*30% == 0`.

Reply:
The main changes we need is in this `MakeSCRestAPICall` function, that we limit the shaders number instead of passing all sharders to `NewTransactionQuery`. Then call the FromAll() to send requests to those selected sharders. We don't need the `fromConsensus`. 

```go
reqSharders := util.Shuffle(copyOf(_config.chain.Sharders))[:reqNum] // util.Shuffle(_config.chain.Sharders) // this shuffle will change the Sharders slice, it's not thread safe. 
tq, err := NewTransactionQuery(reqSharders)
...
```